### ISSUE TITLE
Implement a basic way to solve conflicts

### ISSUE THREAD

**Issue by jbsarrodie** [archi-contribs/archi-modelrepository-plugin#3]
By "basic" I mean: choose globally between "user" or "remote" changes. Of course, this will not fit all needs, but that will mean that it is possible (though limited) to work 100% from within Archi.

So my proposal is to:
* Do the git pull
* Check if there is any conflicts
* If not, ~drink a coffee~ perfect
* If yes, ask user if we should should keep all local (i.e. user) or all remote changes Resolve conflicts accordingly
* ~Drink a beer~ Perfect !

When a user wants to push, in fact the first thing that has to be done is to fech remote changes and then do a local merge (between origin/master and master). This is the goal of [`git pull`](https://git-scm.com/docs/git-pull) which is a shorthand for `git fetch` followed by `git merge`. So no need for a new branch at this stage. If conflicts happen during the merge, then we have to solve them (I haven't looked at how to do it with JGit but I guess we can get conflicts list and then iterate on it) and finally do a `git merge --continue`.

**Comment by jbsarrodie** [comment]
_This basic functionality is now present. On Refresh and Publish merge conflicts are presented to the user in a basic dialog._
* _Press Cancel = abandon the operation and revert to last local commit_
* _Press OK = lose all local commits and take the remote master commit_

_This means that in the first case, a manual conflict resolution needs to be made._

This is not what I had in mind. In my mind...

Imagine both user A and B clone a remote model. Then:
* A adds a new element E1
* A edits some attributes of element E2 which is now E2a
* A commits and share his work

Now the remote model contains E1 and E2v1

In parallel:
* B adds a new element E3
* B edits some attributes of element 2 which is now E2b
* B commits his work and tries to share his work

Whatever B does, at the end the remote model should contains both E1 and E3, but should contains either E2b or E2a, depending on B's choice to keep his or remote changes.

So in my mind when B publish, we should fetch remote changes, try the merge, then we should get the list of conflicts (in this case E2b vs E2a). B should be warned that conflicts exist (maybe with the number of conflicts if easy to get), then asked to choose between his changes and remote one. If he choose his changes, we commit E2b, if he choose remote changes, we commit E2a. Then we push all this.

Is it clearer ?

**Comment by Phillipus** [comment]
But you can't push B's changes if there are conflicts that have to be resolved. As both are using master, the conflict has to be resolved.

**Comment by jbsarrodie** [comment]
I was thinking about using something like `git merge -s ours|theirs` which auto-resolves conflicts.

**Comment by jbsarrodie** [comment]
FWIW, this seems to be doable in [JGit](http://download.eclipse.org/jgit/site/3.7.1.201504261725-r/apidocs/org/eclipse/jgit/api/CheckoutCommand.html#setStage(org.eclipse.jgit.api.CheckoutCommand.Stage)) by doing a checkout of conflicting files on the _ours_ or _theirs_ stage. 

This is equivalent to a [`git checkout --ours|theirs path`](http://gitready.com/advanced/2009/02/25/keep-either-file-in-merge-conflicts.html)

**Comment by Phillipus** [comment]
I can Pull and specify a strategy of ours or theirs. What I need to find out is how to Fetch, then do a dummy Merge, get the conflicts, and then do a ours|theirs merge.

**Comment by jbsarrodie** [comment]
From what I ve read JGit merge stategies OURS and THEIRS are for another purpose (keep either ours or theirs but for the whole tree).

I think the best is to do the merge and do a checkout of each conflicted files with either ours or theirs stage. This way no need for a dumy merge.

**Comment by Phillipus** [comment]
Once a pull is done, there are files in the working tree that have to be resolved.

**Comment by jbsarrodie** [comment]
Yes but once the fetch has been done you can just replace the content of these files through a `git checkout --ours|theirs filepath'  loop. This should be easier than editing files to remove text and conflict marks.

Maybe this should be tested using git on command line...

**Comment by Phillipus** [comment]
Surely you have to select ours or theirs on all conflicting files? Otherwise there may be corrupted model?

**Comment by Phillipus** [comment]
Not sure Checkout is the right thing to do here. This checks out from the Index.

**Comment by jbsarrodie** [comment]
I've just tested using git on command line:
* create a conflicting situation
* do a `git pull` (fail with conflict) : as expected I have a file with conflict marks (<<< === >>>)
* do a `git checkout --ours` : conflicting file is replaced by my version
* do a `git checkout --theirs` : conflicting file is replaced by remote version

**Comment by Phillipus** [comment]
I think I've solved it. Check latest commit. It does a ours or theirs. At the moment you can't choose ours or theirs on individual files, but at least it's a proper merge and not a reset to local or remote.

**Comment by Phillipus** [comment]
You can now choose "ours" or "theirs" for each conflicting file. And view the conflict as text. Next version should have ours and theirs text windows.

**Comment by Phillipus** [comment]
I'd like to mark this as "done" given that we have achieved a MVP and the goal was a "basic" solution. We could open a new issue to track the enhancements such as handling deletions of elements.

**Comment by Phillipus** [comment]
Will open a new issue....

**Comment by jbsarrodie** [comment]
I fully agree. This user story is done.
